perf/perf.go

```diff
@@ -0,0 +1,520 @@
+	return errors.Wrap(err, "can't add fd to epoll")
+}
+
+func (ring *perfEventRing) Close() {
+	runtime.SetFinalizer(ring, nil)
+	unix.Close(ring.fd)
+	unix.Munmap(ring.mmap)
+
+	ring.fd = -1
+	ring.mmap = nil
+}
+
+func (ring *perfEventRing) LoadHead() {
+	ring.reader.loadHead()
+}
+
+// NB: Has to be preceded by a call to LoadHead.
+func (ring *perfEventRing) ReadRecord() (*Record, error) {
+	if ring.fd == -1 {
+		return nil, errors.New("closed")
+	}
+
+	defer ring.reader.writeTail()
+	return readRecord(ring.reader)
+}
+
+// Record contains either a sample or a counter of the
+// number of lost samples.
+type Record struct {
+	// The data submitted via bpf_perf_event_output.
+	// They are padded with 0 to have a 64-bit alignment.
+	// If you are using variable length samples you need to take
+	// this into account.
+	RawSample []byte
+
+	// The number of samples which could not be output, since
+	// the ring buffer was full.
+	LostSamples uint64
+}
+
+func readRecord(rd io.Reader) (*Record, error) {
+	const (
+		perfRecordLost   = 2
+		perfRecordSample = 9
+	)
+
+	var header perfEventHeader
+	err := binary.Read(rd, internal.NativeEndian, &header)
+	if err == io.EOF {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "can't read event header")
+	}
+
+	switch header.Type {
+	case perfRecordLost:
+		lost, err := readLostRecords(rd)
+		if err != nil {
+			return nil, err
+		}
+
+		return &Record{LostSamples: lost}, nil
+
+	case perfRecordSample:
+		raw, err := readRawSample(rd)
+		if err != nil {
+			return nil, err
+		}
+
+		return &Record{RawSample: raw}, nil
+
+	default:
+		return nil, errors.Errorf("unknown event type %d", header.Type)
+	}
+}
+
+func readLostRecords(rd io.Reader) (uint64, error) {
+	var lostHeader struct {
+		ID   uint64
+		Lost uint64
+	}
+
+	err := binary.Read(rd, internal.NativeEndian, &lostHeader)
+	if err != nil {
+		return 0, errors.Wrap(err, "can't read lost records header")
+	}
+
+	return lostHeader.Lost, nil
+}
+
+func readRawSample(rd io.Reader) ([]byte, error) {
+	var size uint32
+	if err := binary.Read(rd, internal.NativeEndian, &size); err != nil {
+		return nil, errors.Wrap(err, "can't read sample size")
+	}
+
+	data := make([]byte, int(size))
+	_, err := io.ReadFull(rd, data)
+	return data, errors.Wrap(err, "can't read sample")
+}
+
+// Reader allows reading bpf_perf_event_output
+// from user space.
+type Reader struct {
+	mu sync.Mutex
+
+	// Closing a PERF_EVENT_ARRAY removes all event fds
+	// stored in it, so we keep a reference alive.
+	array *ebpf.Map
+	rings []*perfEventRing
+
+	epollFd     int
+	epollEvents []unix.EpollEvent
+	epollRings  []*perfEventRing
+	// Eventfds for closing
+	closeFd int
+	// Ensure we only close once
+	closeOnce sync.Once
+}
+
+// ReaderOptions control the behaviour of the user
+// space reader.
+type ReaderOptions struct {
+	// A map of type PerfEventArray.
+	Map *ebpf.Map
+	// Controls the size of the per CPU buffer in bytes.
+	PerCPUBuffer int
```

Comment:
For performance reasons we presumably would want to align this with page size, is there any benefit to allowing arbitrary byte-sized buffers?

From Cilium I'm expecting to basically do a `pageSize := os.Getpagesize()` then pass some multiple of that in here. Not a big deal by any means, but making it easier for users of the library to do the right thing might guide us to just expose a "number of pages" thing here.

Reply:
I think it rounds up to the nearest page multiple. Using number of pages makes the buffer size hard to understand IMO, since page sizes are variable between arches / kernel configs.

Reply:
I think it's reasonable to specify via bytes. Easier to actually manage your application's memory footprint when the units are the standard units.